From b5963613f1918a7199d2bbcc8d1e8eae718b04f4 Mon Sep 17 00:00:00 2001
From: Alec Ari <neotheuser@ymail.com>
Date: Thu, 14 Sep 2023 20:53:05 +0000
Subject: Countless Kconfig tweaks for IPIPE

Disables every problematic option I can think of and enables
several very important options for debugging, or those required
for base RTAI functionality. RTAI may cause kernel panics
so it is a good idea to always have these debugging features on.
The options enabled do not cause an increase in overhead.

Various other fixes and optimizations are included as well,
including CONFIG_DEBUG_FORCE_FUNCTION_ALIGN_64B

DRM drivers (including those that depend on WMI such as
Intel and Nouveau) are now re-enabled. If you experience
long latency delays, it is highly recommended to disable
these drivers as a troubleshooting step.

Signed-off-by: Alec Ari <neotheuser@ymail.com>
---
 Makefile                             |   4 +
 arch/Kconfig                         |   3 +-
 arch/x86/Kconfig                     | 145 +++++++++++++++------------
 arch/x86/Kconfig.debug               |  15 ++-
 arch/x86/events/Kconfig              |   1 +
 arch/x86/kvm/Kconfig                 |   1 +
 certs/Kconfig                        |   1 +
 drivers/acpi/Kconfig                 |  23 +++--
 drivers/acpi/dptf/Kconfig            |   2 +-
 drivers/ata/Kconfig                  |   3 +-
 drivers/bus/Kconfig                  |   1 +
 drivers/char/ipmi/Kconfig            |   2 +-
 drivers/char/tpm/Kconfig             |   2 +-
 drivers/clk/Kconfig                  |   2 +-
 drivers/clocksource/Kconfig          |  20 ++--
 drivers/cpufreq/Kconfig              |   1 +
 drivers/cpuidle/Kconfig              |   1 +
 drivers/crypto/Kconfig               |   1 +
 drivers/devfreq/Kconfig              |   1 +
 drivers/dma/Kconfig                  |   2 +-
 drivers/firmware/Kconfig             |   2 +-
 drivers/firmware/tegra/Kconfig       |   1 +
 drivers/gpu/drm/i915/Kconfig         |   2 +
 drivers/gpu/drm/nouveau/Kconfig      |  12 ++-
 drivers/hv/Kconfig                   |   1 +
 drivers/hwspinlock/Kconfig           |   1 +
 drivers/i2c/Kconfig                  |   2 +-
 drivers/iommu/Kconfig                |   2 +-
 drivers/mailbox/Kconfig              |   1 +
 drivers/memory/Kconfig               |   1 +
 drivers/mfd/Kconfig                  |   3 +-
 drivers/net/ethernet/stmicro/Kconfig |   2 +-
 drivers/of/Kconfig                   |   1 +
 drivers/pci/Kconfig                  |   6 +-
 drivers/pci/hotplug/Kconfig          |   2 +-
 drivers/pci/pcie/Kconfig             |   2 +-
 drivers/perf/Kconfig                 |   2 +-
 drivers/pinctrl/Kconfig              |   1 +
 drivers/power/avs/Kconfig            |   1 +
 drivers/power/reset/Kconfig          |   1 +
 drivers/powercap/Kconfig             |   1 +
 drivers/pwm/Kconfig                  |   1 +
 drivers/ras/Kconfig                  |   1 +
 drivers/regulator/Kconfig            |   1 +
 drivers/remoteproc/Kconfig           |   1 +
 drivers/reset/Kconfig                |   1 +
 drivers/rpmsg/Kconfig                |   1 +
 drivers/scsi/ufs/Kconfig             |   2 +-
 drivers/sfi/Kconfig                  |   1 +
 drivers/soc/Kconfig                  |   1 +
 drivers/usb/chipidea/Kconfig         |   2 +-
 drivers/virt/Kconfig                 |   1 +
 drivers/virtio/Kconfig               |   1 +
 drivers/watchdog/Kconfig             |   1 +
 fs/Kconfig.binfmt                    |   1 +
 include/linux/ipipe.h                |   2 -
 init/Kconfig                         |  34 ++++---
 kernel/gcov/Kconfig                  |   1 +
 kernel/ipipe/Kconfig                 |  45 ++++++---
 kernel/ipipe/Kconfig.debug           |   6 --
 kernel/irq/Kconfig                   |   2 +-
 kernel/power/Kconfig                 |   1 +
 kernel/rcu/Kconfig.debug             |   2 +-
 kernel/trace/Kconfig                 |   1 +
 lib/Kconfig.debug                    |  44 +++++---
 lib/Kconfig.ubsan                    |   1 +
 lib/xz/Kconfig                       |   5 +
 mm/Kconfig                           |   4 +-
 mm/Kconfig.debug                     |   2 +
 net/wireless/Kconfig                 |   4 +-
 samples/Kconfig                      |   2 +-
 security/Kconfig                     |   7 +-
 sound/soc/Kconfig                    |   1 +
 73 files changed, 288 insertions(+), 170 deletions(-)

diff --git a/Makefile b/Makefile
index 4f8f98c7227a..0432676d97af 100644
--- a/Makefile
+++ b/Makefile
@@ -812,6 +812,10 @@ KBUILD_CFLAGS_KERNEL += -ffunction-sections -fdata-sections
 LDFLAGS_vmlinux += --gc-sections
 endif
 
+ifdef CONFIG_DEBUG_FORCE_FUNCTION_ALIGN_64B
+KBUILD_CFLAGS += -falign-functions=64
+endif
+
 # arch Makefile may override CC so keep this after arch Makefile is included
 NOSTDINC_FLAGS += -nostdinc -isystem $(shell $(CC) -print-file-name=include)
 
diff --git a/arch/Kconfig b/arch/Kconfig
index 6e77e795517d..f3646deaa447 100644
--- a/arch/Kconfig
+++ b/arch/Kconfig
@@ -465,7 +465,7 @@ config STACKPROTECTOR
 
 config STACKPROTECTOR_STRONG
 	bool "Strong Stack Protector"
-	depends on STACKPROTECTOR
+	depends on STACKPROTECTOR && !IPIPE
 	depends on $(cc-option,-fstack-protector-strong)
 	default y
 	help
@@ -848,6 +848,7 @@ config ARCH_HAS_REFCOUNT
 
 config REFCOUNT_FULL
 	bool "Perform full reference count validation at the expense of speed"
+	depends on !IPIPE
 	help
 	  Enabling this switches the refcounting infrastructure from a fast
 	  unchecked atomic_t implementation to a fully state checked
diff --git a/arch/x86/Kconfig b/arch/x86/Kconfig
index 1862453e9eac..46b92140ff31 100644
--- a/arch/x86/Kconfig
+++ b/arch/x86/Kconfig
@@ -51,16 +51,16 @@ config X86
 	select ARCH_DISCARD_MEMBLOCK
 	select ARCH_HAS_ACPI_TABLE_UPGRADE	if ACPI
 	select ARCH_HAS_CPU_FINALIZE_INIT
-	select ARCH_HAS_DEBUG_VIRTUAL
+	select ARCH_HAS_DEBUG_VIRTUAL		if !IPIPE
 	select ARCH_HAS_DEVMEM_IS_ALLOWED
 	select ARCH_HAS_ELF_RANDOMIZE
 	select ARCH_HAS_FAST_MULTIPLIER
 	select ARCH_HAS_FILTER_PGPROT
 	select ARCH_HAS_FORTIFY_SOURCE
-	select ARCH_HAS_GCOV_PROFILE_ALL
-	select ARCH_HAS_KCOV			if X86_64
+	select ARCH_HAS_GCOV_PROFILE_ALL	if !IPIPE
+	select ARCH_HAS_KCOV			if X86_64 && !IPIPE
 	select ARCH_HAS_MEMBARRIER_SYNC_CORE
-	select ARCH_HAS_PMEM_API		if X86_64
+	select ARCH_HAS_PMEM_API		if X86_64 && !IPIPE
 	select ARCH_HAS_PTE_SPECIAL
 	select ARCH_HAS_REFCOUNT
 	select ARCH_HAS_UACCESS_FLUSHCACHE	if X86_64
@@ -70,7 +70,7 @@ config X86
 	select ARCH_HAS_STRICT_KERNEL_RWX
 	select ARCH_HAS_STRICT_MODULE_RWX
 	select ARCH_HAS_SYNC_CORE_BEFORE_USERMODE
-	select ARCH_HAS_UBSAN_SANITIZE_ALL
+	select ARCH_HAS_UBSAN_SANITIZE_ALL	if !IPIPE
 	select ARCH_HAS_ZONE_DEVICE		if X86_64
 	select ARCH_HAVE_NMI_SAFE_CMPXCHG
 	select ARCH_MIGHT_HAVE_ACPI_PDC		if ACPI
@@ -78,7 +78,7 @@ config X86
 	select ARCH_MIGHT_HAVE_PC_SERIO
 	select ARCH_SUPPORTS_ACPI
 	select ARCH_SUPPORTS_ATOMIC_RMW
-	select ARCH_SUPPORTS_NUMA_BALANCING	if X86_64
+	select ARCH_SUPPORTS_NUMA_BALANCING	if X86_64 && !IPIPE
 	select ARCH_USE_BUILTIN_BSWAP
 	select ARCH_USE_QUEUED_RWLOCKS
 	select ARCH_USE_QUEUED_SPINLOCKS
@@ -91,14 +91,14 @@ config X86
 	select CLOCKSOURCE_WATCHDOG
 	select DCACHE_WORD_ACCESS
 	select DMA_DIRECT_OPS
-	select EDAC_ATOMIC_SCRUB
-	select EDAC_SUPPORT
+	select EDAC_ATOMIC_SCRUB		if !IPIPE
+	select EDAC_SUPPORT			if !IPIPE
 	select GENERIC_CLOCKEVENTS
 	select GENERIC_CLOCKEVENTS_BROADCAST	if X86_64 || (X86_32 && X86_LOCAL_APIC)
 	select GENERIC_CLOCKEVENTS_MIN_ADJUST
 	select GENERIC_CMOS_UPDATE
 	select GENERIC_CPU_AUTOPROBE
-	select GENERIC_CPU_VULNERABILITIES
+	select GENERIC_CPU_VULNERABILITIES	if !IPIPE
 	select GENERIC_EARLY_IOREMAP
 	select GENERIC_FIND_FIRST_BIT
 	select GENERIC_IOMAP
@@ -113,15 +113,15 @@ config X86
 	select GENERIC_STRNCPY_FROM_USER
 	select GENERIC_STRNLEN_USER
 	select GENERIC_TIME_VSYSCALL
-	select HARDLOCKUP_CHECK_TIMESTAMP	if X86_64
-	select HAVE_ACPI_APEI			if ACPI
-	select HAVE_ACPI_APEI_NMI		if ACPI
+	select HARDLOCKUP_CHECK_TIMESTAMP	if X86_64 && !IPIPE
+	select HAVE_ACPI_APEI			if ACPI && !IPIPE
+	select HAVE_ACPI_APEI_NMI		if ACPI && !IPIPE
 	select HAVE_ALIGNED_STRUCT_PAGE		if SLUB
-	select HAVE_ARCH_AUDITSYSCALL
+	select HAVE_ARCH_AUDITSYSCALL		if !IPIPE
 	select HAVE_ARCH_HUGE_VMAP		if X86_64 || X86_PAE
 	select HAVE_ARCH_JUMP_LABEL		if !IPIPE
-	select HAVE_ARCH_KASAN			if X86_64
-	select HAVE_ARCH_KGDB
+	select HAVE_ARCH_KASAN			if X86_64 && !IPIPE
+	select HAVE_ARCH_KGDB			if !IPIPE
 	select HAVE_ARCH_MMAP_RND_BITS		if MMU
 	select HAVE_ARCH_MMAP_RND_COMPAT_BITS	if MMU && COMPAT
 	select HAVE_ARCH_COMPAT_MMAP_BASES	if MMU && COMPAT
@@ -138,52 +138,47 @@ config X86
 	select HAVE_CONTEXT_TRACKING		if X86_64 && !IPIPE
 	select HAVE_COPY_THREAD_TLS
 	select HAVE_C_RECORDMCOUNT
-	select HAVE_DEBUG_KMEMLEAK
-	select HAVE_DEBUG_STACKOVERFLOW
+	select HAVE_DEBUG_KMEMLEAK		if !IPIPE
+	select HAVE_DEBUG_STACKOVERFLOW		if !IPIPE
 	select HAVE_DMA_CONTIGUOUS
-	select HAVE_DYNAMIC_FTRACE
-	select HAVE_DYNAMIC_FTRACE_WITH_REGS
+	select HAVE_DYNAMIC_FTRACE		if !IPIPE
+	select HAVE_DYNAMIC_FTRACE_WITH_REGS	if !IPIPE
 	select HAVE_EBPF_JIT
 	select HAVE_EFFICIENT_UNALIGNED_ACCESS
 	select HAVE_EXIT_THREAD
 	select HAVE_FENTRY			if X86_64 || DYNAMIC_FTRACE
-	select HAVE_FTRACE_MCOUNT_RECORD
-	select HAVE_FUNCTION_GRAPH_TRACER
-	select HAVE_FUNCTION_TRACER
-	select HAVE_GCC_PLUGINS
+	select HAVE_FTRACE_MCOUNT_RECORD	if !IPIPE
+	select HAVE_FUNCTION_GRAPH_TRACER	if !IPIPE
+	select HAVE_FUNCTION_TRACER		if !IPIPE
+	select HAVE_GCC_PLUGINS			if !IPIPE
 	select HAVE_HW_BREAKPOINT
 	select HAVE_IDE
 	select HAVE_IOREMAP_PROT
 	select HAVE_IRQ_EXIT_ON_IRQ_STACK	if X86_64
 	select HAVE_IRQ_TIME_ACCOUNTING
-	select HAVE_IPIPE_SUPPORT		if X86_64
 	select HAVE_IPIPE_TRACER_SUPPORT
-	select IPIPE_HAVE_HOSTRT if IPIPE
-	select IPIPE_HAVE_SAFE_THREAD_INFO if IPIPE
-	select IPIPE_WANT_PTE_PINNING if IPIPE
-	select IPIPE_HAVE_VM_NOTIFIER if IPIPE
 	select HAVE_KERNEL_BZIP2
 	select HAVE_KERNEL_GZIP
 	select HAVE_KERNEL_LZ4
 	select HAVE_KERNEL_LZMA
 	select HAVE_KERNEL_LZO
 	select HAVE_KERNEL_XZ
-	select HAVE_KPROBES
-	select HAVE_KPROBES_ON_FTRACE
-	select HAVE_FUNCTION_ERROR_INJECTION
-	select HAVE_KRETPROBES
-	select HAVE_KVM
-	select HAVE_LIVEPATCH			if X86_64
+	select HAVE_KPROBES			if !IPIPE
+	select HAVE_KPROBES_ON_FTRACE		if !IPIPE
+	select HAVE_FUNCTION_ERROR_INJECTION	if !IPIPE
+	select HAVE_KRETPROBES			if !IPIPE
+	select HAVE_KVM				if !IPIPE
+	select HAVE_LIVEPATCH			if X86_64 && !IPIPE
 	select HAVE_MEMBLOCK
 	select HAVE_MEMBLOCK_NODE_MAP
 	select HAVE_MIXED_BREAKPOINTS_REGS
 	select HAVE_MOD_ARCH_SPECIFIC
 	select HAVE_NMI
-	select HAVE_OPROFILE
+	select HAVE_OPROFILE			if !IPIPE
 	select HAVE_OPTPROBES
-	select HAVE_PCSPKR_PLATFORM
+	select HAVE_PCSPKR_PLATFORM		if !IPIPE
 	select HAVE_PERF_EVENTS
-	select HAVE_PERF_EVENTS_NMI
+	select HAVE_PERF_EVENTS_NMI		if !IPIPE
 	select HAVE_HARDLOCKUP_DETECTOR_PERF	if PERF_EVENTS && HAVE_PERF_EVENTS_NMI
 	select HAVE_PERF_REGS
 	select HAVE_PERF_USER_STACK_DUMP
@@ -192,12 +187,16 @@ config X86
 	select HAVE_REGS_AND_STACK_ACCESS_API
 	select HAVE_RELIABLE_STACKTRACE		if X86_64 && (UNWINDER_FRAME_POINTER || UNWINDER_ORC) && STACK_VALIDATION
 	select HAVE_STACKPROTECTOR		if CC_HAS_SANE_STACKPROTECTOR
-	select HAVE_STACK_VALIDATION		if X86_64
+	select HAVE_STACK_VALIDATION		if X86_64 && !IPIPE
 	select HAVE_RSEQ
 	select HAVE_SYSCALL_TRACEPOINTS
 	select HAVE_UNSTABLE_SCHED_CLOCK
 	select HAVE_USER_RETURN_NOTIFIER
-	select HOTPLUG_SMT			if SMP
+	select HOTPLUG_SMT			if SMP && !IPIPE
+	# IPIPE should never be disabled
+	select IPIPE
+	select IPIPE_HAVE_HOSTRT if IPIPE
+	select IPIPE_WANT_PTE_PINNING if IPIPE
 	select IRQ_FORCED_THREADING
 	select NEED_SG_DMA_LENGTH
 	select PCI_LOCKLESS_CONFIG
@@ -297,9 +296,11 @@ config NEED_PER_CPU_PAGE_FIRST_CHUNK
 
 config ARCH_HIBERNATION_POSSIBLE
 	def_bool y
+	depends on !IPIPE
 
 config ARCH_SUSPEND_POSSIBLE
 	def_bool y
+	depends on !IPIPE
 
 config ARCH_WANT_HUGE_PMD_SHARE
 	def_bool y
@@ -312,6 +313,7 @@ config ZONE_DMA32
 
 config AUDIT_ARCH
 	def_bool y if X86_64
+	depends on !IPIPE
 
 config ARCH_SUPPORTS_OPTIMIZED_INLINING
 	def_bool y
@@ -429,7 +431,7 @@ config X86_X2APIC
 config X86_MPPARSE
 	bool "Enable MPS table" if ACPI || SFI
 	default y
-	depends on X86_LOCAL_APIC
+	depends on X86_LOCAL_APIC && !IPIPE
 	---help---
 	  For old smp systems that do not have proper acpi support. Newer systems
 	  (esp with 64bit cpus) with acpi support, MADT and DSDT will override it
@@ -441,6 +443,7 @@ config GOLDFISH
 config RETPOLINE
 	bool "Avoid speculative indirect branches in kernel"
 	default y
+	depends on !IPIPE
 	select STACK_VALIDATION if HAVE_STACK_VALIDATION
 	help
 	  Compile kernel with the retpoline compiler options to guard against
@@ -492,6 +495,7 @@ endif
 if X86_64
 config X86_EXTENDED_PLATFORM
 	bool "Support for extended (non-PC) x86 platforms"
+	depends on !IPIPE
 	default y
 	---help---
 	  If you disable this option then the kernel will only support
@@ -614,7 +618,7 @@ config X86_INTEL_QUARK
 
 config X86_INTEL_LPSS
 	bool "Intel Low Power Subsystem Support"
-	depends on X86 && ACPI
+	depends on X86 && ACPI && !IPIPE
 	select COMMON_CLK
 	select PINCTRL
 	select IOSF_MBI
@@ -626,7 +630,7 @@ config X86_INTEL_LPSS
 
 config X86_AMD_PLATFORM_DEVICE
 	bool "AMD ACPI2Platform devices support"
-	depends on ACPI
+	depends on ACPI && !IPIPE
 	select COMMON_CLK
 	select PINCTRL
 	---help---
@@ -741,6 +745,7 @@ config SCHED_OMIT_FRAME_POINTER
 
 menuconfig HYPERVISOR_GUEST
 	bool "Linux guest support"
+	depends on !IPIPE
 	---help---
 	  Say Y here to enable options for running Linux under various hyper-
 	  visors. This option enables basic hypervisor detection and platform
@@ -753,7 +758,6 @@ if HYPERVISOR_GUEST
 
 config PARAVIRT
 	bool "Enable paravirtualization code"
-	depends on !IPIPE
 	---help---
 	  This changes the kernel so it can modify itself when it is run
 	  under a hypervisor, potentially improving performance significantly
@@ -893,7 +897,7 @@ config GART_IOMMU
 	bool "Old AMD GART IOMMU support"
 	select IOMMU_HELPER
 	select SWIOTLB
-	depends on X86_64 && PCI && AMD_NB
+	depends on X86_64 && PCI && AMD_NB && !IPIPE
 	---help---
 	  Provides a driver for older AMD Athlon64/Opteron/Turion/Sempron
 	  GART based hardware IOMMUs.
@@ -915,7 +919,7 @@ config CALGARY_IOMMU
 	bool "IBM Calgary IOMMU support"
 	select IOMMU_HELPER
 	select SWIOTLB
-	depends on X86_64 && PCI
+	depends on X86_64 && PCI && !IPIPE
 	---help---
 	  Support for hardware IOMMUs in IBM's xSeries x366 and x460
 	  systems. Needed to run systems with more than 3GB of memory
@@ -979,8 +983,9 @@ config NR_CPUS_RANGE_END
 config NR_CPUS_RANGE_END
 	int
 	depends on X86_64
-	default 8192 if  SMP && ( MAXSMP ||  CPUMASK_OFFSTACK)
-	default  512 if  SMP && (!MAXSMP && !CPUMASK_OFFSTACK)
+	default 8192 if  SMP && ( MAXSMP ||  CPUMASK_OFFSTACK) && !IPIPE
+	default  512 if  SMP && (!MAXSMP && !CPUMASK_OFFSTACK) && !IPIPE
+	default   32 if  SMP && IPIPE
 	default    1 if !SMP
 
 config NR_CPUS_DEFAULT
@@ -995,6 +1000,7 @@ config NR_CPUS_DEFAULT
 	depends on X86_64
 	default 8192 if  MAXSMP
 	default   64 if  SMP
+	default    8 if  IPIPE
 	default    1 if !SMP
 
 config NR_CPUS
@@ -1007,16 +1013,19 @@ config NR_CPUS
 	  supported value is 8192, otherwise the maximum value is 512.  The
 	  minimum value which makes sense is 2.
 
-	  This is purely to save memory: each supported CPU adds about 8KB
-	  to the kernel image.
+	  For IPIPE-enabled platforms, large values causes serious problems.
+	  The max limit in this case is 16.
+
+	  Each supported CPU adds approximately 8KB to the kernel image.
 
 config SCHED_SMT
 	def_bool y if SMP
+	depends on !IPIPE
 
 config SCHED_MC
 	def_bool y
 	prompt "Multi-core scheduler support"
-	depends on SMP
+	depends on SMP && !IPIPE
 	---help---
 	  Multi-core scheduler support improves the CPU scheduler's decision
 	  making when dealing with multi-core CPU chips at a cost of slightly
@@ -1110,6 +1119,7 @@ config X86_REROUTE_FOR_BROKEN_BOOT_IRQS
 
 config X86_MCE
 	bool "Machine Check / overheating reporting"
+	depends on !IPIPE
 	select GENERIC_ALLOCATOR
 	default y
 	---help---
@@ -1221,7 +1231,7 @@ config X86_ESPFIX64
 config X86_VSYSCALL_EMULATION
        bool "Enable vsyscall emulation" if EXPERT
        default y
-       depends on X86_64
+       depends on X86_64 && !IPIPE
        ---help---
 	 This enables emulation of the legacy vsyscall page.  Disabling
 	 it is roughly equivalent to booting with vsyscall=none, except
@@ -1254,6 +1264,7 @@ config TOSHIBA
 
 config I8K
 	tristate "Dell i8k legacy laptop support"
+	depends on !IPIPE
 	select HWMON
 	select SENSORS_DELL_SMM
 	---help---
@@ -1468,7 +1479,7 @@ config X86_5LEVEL
 	bool "Enable 5-level page tables support"
 	select DYNAMIC_MEMORY_LAYOUT
 	select SPARSEMEM_VMEMMAP
-	depends on X86_64
+	depends on X86_64 && !IPIPE
 	---help---
 	  5-level paging enables access to larger address space:
 	  upto 128 PiB of virtual address space and 4 PiB of
@@ -1498,7 +1509,7 @@ config ARCH_HAS_MEM_ENCRYPT
 
 config AMD_MEM_ENCRYPT
 	bool "AMD Secure Memory Encryption (SME) support"
-	depends on X86_64 && CPU_SUP_AMD
+	depends on X86_64 && CPU_SUP_AMD && !IPIPE
 	select DYNAMIC_PHYSICAL_MASK
 	select ARCH_USE_MEMREMAP_PROT
 	---help---
@@ -1524,6 +1535,7 @@ config NUMA
 	bool "Numa Memory Allocation and Scheduler Support"
 	depends on SMP
 	depends on X86_64 || (X86_32 && HIGHMEM64G && X86_BIGSMP)
+	depends on !IPIPE
 	default y if X86_BIGSMP
 	---help---
 	  Enable NUMA (Non Uniform Memory Access) support.
@@ -1639,8 +1651,7 @@ config X86_PMEM_LEGACY_DEVICE
 
 config X86_PMEM_LEGACY
 	tristate "Support non-standard NVDIMMs and ADR protected memory"
-	depends on PHYS_ADDR_T_64BIT
-	depends on BLK_DEV
+	depends on PHYS_ADDR_T_64BIT && BLK_DEV && !IPIPE
 	select X86_PMEM_LEGACY_DEVICE
 	select LIBNVDIMM
 	help
@@ -1842,6 +1853,7 @@ config ARCH_RANDOM
 
 config X86_SMAP
 	def_bool y
+	depends on !IPIPE
 	prompt "Supervisor Mode Access Prevention" if EXPERT
 	---help---
 	  Supervisor Mode Access Prevention (SMAP) is a security
@@ -1853,7 +1865,7 @@ config X86_SMAP
 
 config X86_INTEL_UMIP
 	def_bool y
-	depends on CPU_SUP_INTEL
+	depends on CPU_SUP_INTEL && !IPIPE
 	prompt "Intel User Mode Instruction Prevention" if EXPERT
 	---help---
 	  The User Mode Instruction Prevention (UMIP) is a security
@@ -1871,7 +1883,7 @@ config X86_INTEL_MPX
 	prompt "Intel MPX (Memory Protection Extensions)"
 	def_bool n
 	# Note: only available in 64-bit mode due to VMA flags shortage
-	depends on CPU_SUP_INTEL && X86_64
+	depends on CPU_SUP_INTEL && X86_64 && !IPIPE
 	select ARCH_USES_HIGH_VMA_FLAGS
 	---help---
 	  MPX provides hardware features that can be used in
@@ -1899,7 +1911,7 @@ config X86_INTEL_MEMORY_PROTECTION_KEYS
 	prompt "Intel Memory Protection Keys"
 	def_bool y
 	# Note: only available in 64-bit mode
-	depends on CPU_SUP_INTEL && X86_64
+	depends on CPU_SUP_INTEL && X86_64 && !IPIPE
 	select ARCH_USES_HIGH_VMA_FLAGS
 	select ARCH_HAS_PKEYS
 	---help---
@@ -1958,7 +1970,7 @@ endchoice
 
 config EFI
 	bool "EFI runtime service support"
-	depends on ACPI
+	depends on ACPI && !IPIPE
 	select UCS2_STRING
 	select EFI_RUNTIME_WRAPPERS
 	select ARCH_USE_MEMREMAP_PROT
@@ -2017,6 +2029,7 @@ source kernel/Kconfig.hz
 
 config KEXEC
 	bool "kexec system call"
+	depends on !IPIPE
 	select KEXEC_CORE
 	---help---
 	  kexec is a system call that implements the ability to shutdown your
@@ -2039,6 +2052,7 @@ config KEXEC_FILE
 	depends on X86_64
 	depends on CRYPTO=y
 	depends on CRYPTO_SHA256=y
+	depends on !IPIPE
 	---help---
 	  This is new version of kexec system call. This system call is
 	  file based and takes file descriptors as system call argument
@@ -2070,6 +2084,7 @@ config KEXEC_BZIMAGE_VERIFY_SIG
 config CRASH_DUMP
 	bool "kernel crash dumps"
 	depends on X86_64 || (X86_32 && HIGHMEM)
+	depends on !IPIPE
 	---help---
 	  Generate crash dump after being started by kexec.
 	  This should be normally only set in special crash dump kernels
@@ -2131,6 +2146,7 @@ config PHYSICAL_START
 
 config RELOCATABLE
 	bool "Build a relocatable kernel"
+	depends on !IPIPE
 	default y
 	---help---
 	  This builds a kernel image that retains relocation information
@@ -2254,7 +2270,7 @@ config RANDOMIZE_MEMORY_PHYSICAL_PADDING
 
 config HOTPLUG_CPU
 	def_bool y
-	depends on SMP
+	depends on SMP && !IPIPE
 
 config BOOTPARAM_HOTPLUG_CPU0
 	bool "Set default setting of cpu0_hotpluggable"
@@ -2328,7 +2344,7 @@ config COMPAT_VDSO
 choice
 	prompt "vsyscall table for legacy applications"
 	depends on X86_64
-	default LEGACY_VSYSCALL_EMULATE
+	default LEGACY_VSYSCALL_NONE
 	help
 	  Legacy user code that does not know how to find the vDSO expects
 	  to be able to issue three syscalls by calling fixed addresses in
@@ -2342,10 +2358,11 @@ choice
 	  static binaries, you can say None without a performance penalty
 	  to improve security.
 
-	  If unsure, select "Emulate".
+	  If unsure, select "None".
 
 	config LEGACY_VSYSCALL_EMULATE
 		bool "Emulate"
+		depends on !IPIPE
 		help
 		  The kernel traps and emulates calls into the fixed
 		  vsyscall address mapping. This makes the mapping
@@ -2410,6 +2427,7 @@ config CMDLINE_OVERRIDE
 
 config MODIFY_LDT_SYSCALL
 	bool "Enable the LDT (local descriptor table)" if EXPERT
+	depends on !IPIPE
 	default y
 	---help---
 	  Linux can allow user programs to install a per-process x86
@@ -2434,7 +2452,7 @@ config ARCH_HAS_ADD_PAGES
 
 config ARCH_ENABLE_MEMORY_HOTPLUG
 	def_bool y
-	depends on X86_64 || (X86_32 && HIGHMEM)
+	depends on X86_64 && !IPIPE || (X86_32 && HIGHMEM)
 
 config ARCH_ENABLE_MEMORY_HOTREMOVE
 	def_bool y
@@ -2918,6 +2936,7 @@ endmenu
 
 
 menu "Binary Emulations"
+	depends on !IPIPE
 
 config IA32_EMULATION
 	bool "IA32 Emulation"
diff --git a/arch/x86/Kconfig.debug b/arch/x86/Kconfig.debug
index 687cd1a213d5..9c11ac3a3b7f 100644
--- a/arch/x86/Kconfig.debug
+++ b/arch/x86/Kconfig.debug
@@ -211,10 +211,10 @@ config IO_DELAY_TYPE_NONE
 
 choice
 	prompt "IO delay type"
-	default IO_DELAY_0X80
+	default IO_DELAY_NONE
 
 config IO_DELAY_0X80
-	bool "port 0x80 based port-IO delay [recommended]"
+	bool "port 0x80 based port-IO delay"
 	---help---
 	  This is the traditional Linux IO delay used for in/out_p.
 	  It is the most tested hence safest selection here.
@@ -232,7 +232,7 @@ config IO_DELAY_UDELAY
 	  while not having any side-effect on the IO port space.
 
 config IO_DELAY_NONE
-	bool "no port-IO delay"
+	bool "no port-IO delay [recommended]"
 	---help---
 	  No port-IO delay. Will break on old boxes that require port-IO
 	  delay for certain operations. Should work on most new machines.
@@ -339,7 +339,7 @@ config X86_DEBUG_FPU
 
 config PUNIT_ATOM_DEBUG
 	tristate "ATOM Punit debug driver"
-	depends on PCI
+	depends on PCI && !IPIPE
 	select DEBUG_FS
 	select IOSF_MBI
 	---help---
@@ -351,8 +351,7 @@ config PUNIT_ATOM_DEBUG
 
 choice
 	prompt "Choose kernel unwinder"
-	default UNWINDER_ORC if X86_64
-	default UNWINDER_FRAME_POINTER if X86_32
+	default UNWINDER_GUESS
 	---help---
 	  This determines which method will be used for unwinding kernel stack
 	  traces for panics, oopses, bugs, warnings, perf, /proc/<pid>/stack,
@@ -360,7 +359,7 @@ choice
 
 config UNWINDER_ORC
 	bool "ORC unwinder"
-	depends on X86_64
+	depends on X86_64 && !IPIPE
 	select STACK_VALIDATION
 	---help---
 	  This option enables the ORC (Oops Rewind Capability) unwinder for
@@ -376,6 +375,7 @@ config UNWINDER_ORC
 
 config UNWINDER_FRAME_POINTER
 	bool "Frame pointer unwinder"
+	depends on !IPIPE
 	select FRAME_POINTER
 	---help---
 	  This option enables the frame pointer unwinder for unwinding kernel
@@ -391,7 +391,6 @@ config UNWINDER_FRAME_POINTER
 
 config UNWINDER_GUESS
 	bool "Guess unwinder"
-	depends on EXPERT
 	depends on !STACKDEPOT
 	---help---
 	  This option enables the "guess" unwinder for unwinding kernel stack
diff --git a/arch/x86/events/Kconfig b/arch/x86/events/Kconfig
index 9a7a1446cb3a..ddcb4d1ecb9a 100644
--- a/arch/x86/events/Kconfig
+++ b/arch/x86/events/Kconfig
@@ -1,5 +1,6 @@
 # SPDX-License-Identifier: GPL-2.0
 menu "Performance monitoring"
+	depends on !IPIPE
 
 config PERF_EVENTS_INTEL_UNCORE
 	tristate "Intel uncore performance events"
diff --git a/arch/x86/kvm/Kconfig b/arch/x86/kvm/Kconfig
index 1bbec387d289..7a2bb7476279 100644
--- a/arch/x86/kvm/Kconfig
+++ b/arch/x86/kvm/Kconfig
@@ -8,6 +8,7 @@ source "virt/kvm/Kconfig"
 menuconfig VIRTUALIZATION
 	bool "Virtualization"
 	depends on HAVE_KVM || X86
+	depends on !IPIPE
 	default y
 	---help---
 	  Say Y here to get to see options for using your Linux host to run other
diff --git a/certs/Kconfig b/certs/Kconfig
index c94e93d8bccf..375089e37f39 100644
--- a/certs/Kconfig
+++ b/certs/Kconfig
@@ -1,5 +1,6 @@
 # SPDX-License-Identifier: GPL-2.0
 menu "Certificates for signature checking"
+depends on !IPIPE
 
 config MODULE_SIG_KEY
 	string "File name or PKCS#11 URI of module signing key"
diff --git a/drivers/acpi/Kconfig b/drivers/acpi/Kconfig
index dd1eea90f67f..5f20cdee38a3 100644
--- a/drivers/acpi/Kconfig
+++ b/drivers/acpi/Kconfig
@@ -62,6 +62,7 @@ config ACPI_CCA_REQUIRED
 
 config ACPI_DEBUGGER
 	bool "AML debugger interface"
+	depends on !IPIPE
 	select ACPI_DEBUG
 	help
 	  Enable in-kernel debugging of AML facilities: statistics,
@@ -82,6 +83,7 @@ endif
 
 config ACPI_SPCR_TABLE
 	bool "ACPI Serial Port Console Redirection Support"
+	depends on !IPIPE
 	default y if X86
 	help
 	  Enable support for Serial Port Console Redirection (SPCR) Table.
@@ -138,6 +140,7 @@ config ACPI_REV_OVERRIDE_POSSIBLE
 
 config ACPI_EC_DEBUGFS
 	tristate "EC read/write access through /sys/kernel/debug/ec"
+	depends on !IPIPE
 	default n
 	help
 	  Say N to disable Embedded Controller /sys/kernel/debug interface
@@ -156,7 +159,7 @@ config ACPI_EC_DEBUGFS
 
 config ACPI_AC
 	tristate "AC Adapter"
-	depends on X86
+	depends on X86 && !IPIPE
 	select POWER_SUPPLY
 	default y
 	help
@@ -169,7 +172,7 @@ config ACPI_AC
 
 config ACPI_BATTERY
 	tristate "Battery"
-	depends on X86
+	depends on X86 && !IPIPE
 	select POWER_SUPPLY
 	default y
 	help
@@ -234,6 +237,7 @@ config ACPI_TAD
 
 config ACPI_DOCK
 	bool "Dock"
+	depends on !IPIPE
 	help
 	  This driver supports ACPI-controlled docking stations and removable
 	  drive bays such as the IBM Ultrabay and the Dell Module Bay.
@@ -244,7 +248,7 @@ config ACPI_CPU_FREQ_PSS
 
 config ACPI_PROCESSOR_CSTATE
 	def_bool y
-	depends on IA64 || X86
+	depends on IA64 || (X86 && !IPIPE)
 
 config ACPI_PROCESSOR_IDLE
 	bool
@@ -268,6 +272,7 @@ config ACPI_CPPC_LIB
 
 config ACPI_PROCESSOR
 	tristate "Processor"
+	depends on !IPIPE
 	depends on X86 || IA64 || ARM64
 	select ACPI_PROCESSOR_IDLE
 	select ACPI_CPU_FREQ_PSS if X86 || IA64
@@ -361,6 +366,7 @@ config ACPI_TABLE_UPGRADE
 
 config ACPI_DEBUG
 	bool "Debug Statements"
+	depends on !IPIPE
 	default n
 	help
 	  The ACPI subsystem can produce debug output.  Saying Y enables this
@@ -373,7 +379,7 @@ config ACPI_DEBUG
 
 config ACPI_PCI_SLOT
 	bool "PCI slot detection driver"
-	depends on SYSFS
+	depends on SYSFS && !IPIPE
 	default n
 	help
 	  This driver creates entries in /sys/bus/pci/slots/ for all PCI
@@ -383,7 +389,7 @@ config ACPI_PCI_SLOT
 
 config ACPI_CONTAINER
 	bool "Container and Module Devices"
-	default (ACPI_HOTPLUG_MEMORY || ACPI_HOTPLUG_CPU)
+	depends on (ACPI_HOTPLUG_MEMORY || ACPI_HOTPLUG_CPU)
 	help
 	  This driver supports ACPI Container and Module devices (IDs
 	  ACPI0004, PNP0A05, and PNP0A06).
@@ -417,7 +423,7 @@ config ACPI_HOTPLUG_IOAPIC
 
 config ACPI_SBS
 	tristate "Smart Battery System"
-	depends on X86
+	depends on X86 && !IPIPE
 	select POWER_SUPPLY
 	help
 	  This driver supports the Smart Battery System, another
@@ -428,6 +434,7 @@ config ACPI_SBS
 
 config ACPI_HED
 	tristate "Hardware Error Device"
+	depends on !IPIPE
 	help
 	  This driver supports the Hardware Error Device (PNP0C33),
 	  which is used to report some hardware errors notified via
@@ -435,7 +442,7 @@ config ACPI_HED
 
 config ACPI_CUSTOM_METHOD
 	tristate "Allow ACPI methods to be inserted/replaced at run time"
-	depends on DEBUG_FS
+	depends on DEBUG_FS && !IPIPE
 	default n
 	help
 	  This debug facility allows ACPI AML methods to be inserted and/or
@@ -500,6 +507,7 @@ config ACPI_EXTLOG
 
 menuconfig PMIC_OPREGION
 	bool "PMIC (Power Management Integrated Circuit) operation region support"
+	depends on !IPIPE
 	help
 	  Select this option to enable support for ACPI operation
 	  region of the PMIC chip. The operation region can be used
@@ -541,6 +549,7 @@ endif
 
 config ACPI_CONFIGFS
 	tristate "ACPI configfs support"
+	depends on !IPIPE
 	select CONFIGFS_FS
 	help
 	  Select this option to enable support for ACPI configuration from
diff --git a/drivers/acpi/dptf/Kconfig b/drivers/acpi/dptf/Kconfig
index 90a2fd979282..bc4189889638 100644
--- a/drivers/acpi/dptf/Kconfig
+++ b/drivers/acpi/dptf/Kconfig
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 config DPTF_POWER
 	tristate "DPTF Platform Power Participant"
-	depends on X86
+	depends on X86 && !IPIPE
 	help
 	  This driver adds support for Dynamic Platform and Thermal Framework
 	  (DPTF) Platform Power Participant device (INT3407) support.
diff --git a/drivers/ata/Kconfig b/drivers/ata/Kconfig
index 99698d7fe585..567ed495daec 100644
--- a/drivers/ata/Kconfig
+++ b/drivers/ata/Kconfig
@@ -94,7 +94,8 @@ config SATA_AHCI
 
 config SATA_MOBILE_LPM_POLICY
 	int "Default SATA Link Power Management policy for mobile chipsets"
-	range 0 4
+	range 0 4 if !IPIPE
+	range 1 1 if IPIPE
 	default 0
 	depends on SATA_AHCI
 	help
diff --git a/drivers/bus/Kconfig b/drivers/bus/Kconfig
index 1851112ccc29..d405ccd607e6 100644
--- a/drivers/bus/Kconfig
+++ b/drivers/bus/Kconfig
@@ -4,6 +4,7 @@
 #
 
 menu "Bus devices"
+	depends on !IPIPE
 
 config ARM_CCI
 	bool
diff --git a/drivers/char/ipmi/Kconfig b/drivers/char/ipmi/Kconfig
index c108441882cc..e115984e8b2d 100644
--- a/drivers/char/ipmi/Kconfig
+++ b/drivers/char/ipmi/Kconfig
@@ -4,7 +4,7 @@
 
 menuconfig IPMI_HANDLER
        tristate 'IPMI top-level message handler'
-       depends on HAS_IOMEM
+       depends on HAS_IOMEM && !IPIPE
        select IPMI_DMI_DECODE if DMI
        help
          This enables the central IPMI message handler, required for IPMI
diff --git a/drivers/char/tpm/Kconfig b/drivers/char/tpm/Kconfig
index 18c81cbe4704..7944a15138e7 100644
--- a/drivers/char/tpm/Kconfig
+++ b/drivers/char/tpm/Kconfig
@@ -4,7 +4,7 @@
 
 menuconfig TCG_TPM
 	tristate "TPM Hardware Support"
-	depends on HAS_IOMEM
+	depends on HAS_IOMEM && !IPIPE
 	select SECURITYFS
 	select CRYPTO
 	select CRYPTO_HASH_INFO
diff --git a/drivers/clk/Kconfig b/drivers/clk/Kconfig
index ffe81449ce24..d57ff707e01c 100644
--- a/drivers/clk/Kconfig
+++ b/drivers/clk/Kconfig
@@ -20,7 +20,7 @@ config COMMON_CLK
 	  this option.
 
 menu "Common Clock Framework"
-	depends on COMMON_CLK
+	depends on COMMON_CLK && !IPIPE
 
 config COMMON_CLK_WM831X
 	tristate "Clock driver for WM831x/2x PMICs"
diff --git a/drivers/clocksource/Kconfig b/drivers/clocksource/Kconfig
index 06504384c376..97ee50c93233 100644
--- a/drivers/clocksource/Kconfig
+++ b/drivers/clocksource/Kconfig
@@ -1,5 +1,14 @@
+config CLKEVT_I8253
+	bool
+
+config I8253_LOCK
+	bool
+
+config CLKBLD_I8253
+	def_bool y if CLKSRC_I8253 || CLKEVT_I8253 || I8253_LOCK
+
 menu "Clock Source drivers"
-	depends on GENERIC_CLOCKEVENTS
+	depends on GENERIC_CLOCKEVENTS && !IPIPE
 
 config TIMER_OF
 	bool
@@ -15,19 +24,10 @@ config TIMER_PROBE
 config CLKSRC_I8253
 	bool
 
-config CLKEVT_I8253
-	bool
-
-config I8253_LOCK
-	bool
-
 config OMAP_DM_TIMER
 	bool
 	select TIMER_OF
 
-config CLKBLD_I8253
-	def_bool y if CLKSRC_I8253 || CLKEVT_I8253 || I8253_LOCK
-
 config CLKSRC_MMIO
 	bool
 
diff --git a/drivers/cpufreq/Kconfig b/drivers/cpufreq/Kconfig
index 608af20a3494..a618e6a9e856 100644
--- a/drivers/cpufreq/Kconfig
+++ b/drivers/cpufreq/Kconfig
@@ -1,4 +1,5 @@
 menu "CPU Frequency scaling"
+	depends on !IPIPE
 
 config CPU_FREQ
 	bool "CPU Frequency scaling"
diff --git a/drivers/cpuidle/Kconfig b/drivers/cpuidle/Kconfig
index 7e48eb5bf0a7..c144773419a4 100644
--- a/drivers/cpuidle/Kconfig
+++ b/drivers/cpuidle/Kconfig
@@ -1,4 +1,5 @@
 menu "CPU Idle"
+	depends on !IPIPE
 
 config CPU_IDLE
 	bool "CPU idle PM support"
diff --git a/drivers/crypto/Kconfig b/drivers/crypto/Kconfig
index a825b6444459..fb90d2f452ae 100644
--- a/drivers/crypto/Kconfig
+++ b/drivers/crypto/Kconfig
@@ -1,6 +1,7 @@
 
 menuconfig CRYPTO_HW
 	bool "Hardware crypto devices"
+	depends on !IPIPE
 	default y
 	---help---
 	  Say Y here to get to see options for hardware crypto devices and
diff --git a/drivers/devfreq/Kconfig b/drivers/devfreq/Kconfig
index 4c4ec68b0566..dcc12c2f8448 100644
--- a/drivers/devfreq/Kconfig
+++ b/drivers/devfreq/Kconfig
@@ -1,5 +1,6 @@
 menuconfig PM_DEVFREQ
 	bool "Generic Dynamic Voltage and Frequency Scaling (DVFS) support"
+	depends on !IPIPE
 	select SRCU
 	select PM_OPP
 	help
diff --git a/drivers/dma/Kconfig b/drivers/dma/Kconfig
index e5f31af65aab..b0d2bb5e9e97 100644
--- a/drivers/dma/Kconfig
+++ b/drivers/dma/Kconfig
@@ -4,7 +4,7 @@
 
 menuconfig DMADEVICES
 	bool "DMA Engine support"
-	depends on HAS_DMA
+	depends on HAS_DMA && !IPIPE
 	help
 	  DMA engines can do asynchronous data transfers without
 	  involving the host CPU.  Currently, this framework can be
diff --git a/drivers/firmware/Kconfig b/drivers/firmware/Kconfig
index 1c419e4cea83..de1cadf89bfd 100644
--- a/drivers/firmware/Kconfig
+++ b/drivers/firmware/Kconfig
@@ -159,7 +159,7 @@ config DELL_RBU
 
 config DCDBAS
 	tristate "Dell Systems Management Base Driver"
-	depends on X86
+	depends on X86 && !IPIPE
 	help
 	  The Dell Systems Management Base Driver provides a sysfs interface
 	  for systems management software to perform System Management
diff --git a/drivers/firmware/tegra/Kconfig b/drivers/firmware/tegra/Kconfig
index ff2730d5c468..4cc867d3fe25 100644
--- a/drivers/firmware/tegra/Kconfig
+++ b/drivers/firmware/tegra/Kconfig
@@ -1,4 +1,5 @@
 menu "Tegra firmware driver"
+	depends on !IPIPE
 
 config TEGRA_IVC
 	bool "Tegra IVC protocol"
diff --git a/drivers/gpu/drm/i915/Kconfig b/drivers/gpu/drm/i915/Kconfig
index 33a458b7f1fc..1fd23981de68 100644
--- a/drivers/gpu/drm/i915/Kconfig
+++ b/drivers/gpu/drm/i915/Kconfig
@@ -99,6 +99,7 @@ config DRM_I915_GVT
         bool "Enable Intel GVT-g graphics virtualization host support"
         depends on DRM_I915
         depends on 64BIT
+        depends on !IPIPE
         default n
         help
 	  Choose this option if you want to enable Intel GVT-g graphics
@@ -131,5 +132,6 @@ config DRM_I915_GVT_KVMGT
 menu "drm/i915 Debugging"
 depends on DRM_I915
 depends on EXPERT
+depends on !IPIPE
 source drivers/gpu/drm/i915/Kconfig.debug
 endmenu
diff --git a/drivers/gpu/drm/nouveau/Kconfig b/drivers/gpu/drm/nouveau/Kconfig
index 00d9d77f583a..b87fb75c678b 100644
--- a/drivers/gpu/drm/nouveau/Kconfig
+++ b/drivers/gpu/drm/nouveau/Kconfig
@@ -42,14 +42,15 @@ config NOUVEAU_PLATFORM_DRIVER
 config NOUVEAU_DEBUG
 	int "Maximum debug level"
 	depends on DRM_NOUVEAU
-	range 0 7
-	default 5
+	range 0 7 if !IPIPE
+	range 0 2 if IPIPE
+	default 2
 	help
 	  Selects the maximum debug level to compile support for.
 
 	  0 - fatal
 	  1 - error
-	  2 - warning
+	  2 - warning (maximum if IPIPE enabled)
 	  3 - info
 	  4 - debug
 	  5 - trace (recommended)
@@ -62,8 +63,9 @@ config NOUVEAU_DEBUG
 config NOUVEAU_DEBUG_DEFAULT
 	int "Default debug level"
 	depends on DRM_NOUVEAU
-	range 0 7
-	default 3
+	range 0 7 if !IPIPE
+	range 0 2 if IPIPE
+	default 1
 	help
 	  Selects the default debug level
 
diff --git a/drivers/hv/Kconfig b/drivers/hv/Kconfig
index 1c1a2514d6f3..4280cc29372d 100644
--- a/drivers/hv/Kconfig
+++ b/drivers/hv/Kconfig
@@ -1,6 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 
 menu "Microsoft Hyper-V guest support"
+	depends on !IPIPE
 
 config HYPERV
 	tristate "Microsoft Hyper-V client drivers"
diff --git a/drivers/hwspinlock/Kconfig b/drivers/hwspinlock/Kconfig
index e895d29500ee..15e95a663c21 100644
--- a/drivers/hwspinlock/Kconfig
+++ b/drivers/hwspinlock/Kconfig
@@ -5,6 +5,7 @@
 
 menuconfig HWSPINLOCK
 	bool "Hardware Spinlock drivers"
+	depends on !IPIPE
 
 config HWSPINLOCK_OMAP
 	tristate "OMAP Hardware Spinlock device"
diff --git a/drivers/i2c/Kconfig b/drivers/i2c/Kconfig
index efc3354d60ae..f64ed101a997 100644
--- a/drivers/i2c/Kconfig
+++ b/drivers/i2c/Kconfig
@@ -26,7 +26,7 @@ config I2C
 
 config ACPI_I2C_OPREGION
 	bool "ACPI I2C Operation region support"
-	depends on I2C=y && ACPI
+	depends on I2C=y && ACPI && !IPIPE
 	default y
 	help
 	  Say Y here if you want to enable ACPI I2C operation region support.
diff --git a/drivers/iommu/Kconfig b/drivers/iommu/Kconfig
index c60395b7470f..921f509b1dfa 100644
--- a/drivers/iommu/Kconfig
+++ b/drivers/iommu/Kconfig
@@ -4,7 +4,7 @@ config IOMMU_API
 
 menuconfig IOMMU_SUPPORT
 	bool "IOMMU Hardware Support"
-	depends on MMU
+	depends on MMU && !IPIPE
 	default y
 	---help---
 	  Say Y here if you want to compile device drivers for IO Memory
diff --git a/drivers/mailbox/Kconfig b/drivers/mailbox/Kconfig
index 841c005d8ebb..a10865c6029e 100644
--- a/drivers/mailbox/Kconfig
+++ b/drivers/mailbox/Kconfig
@@ -1,5 +1,6 @@
 menuconfig MAILBOX
 	bool "Mailbox Hardware Support"
+	depends on !IPIPE
 	help
 	  Mailbox is a framework to control hardware communication between
 	  on-chip processors through queued messages and interrupt driven
diff --git a/drivers/memory/Kconfig b/drivers/memory/Kconfig
index 63389f075f1d..6b7bc8a61837 100644
--- a/drivers/memory/Kconfig
+++ b/drivers/memory/Kconfig
@@ -3,6 +3,7 @@
 #
 
 menuconfig MEMORY
+	depends on !IPIPE
 	bool "Memory Controller drivers"
 
 if MEMORY
diff --git a/drivers/mfd/Kconfig b/drivers/mfd/Kconfig
index dd938a5d0409..0692e2ddb474 100644
--- a/drivers/mfd/Kconfig
+++ b/drivers/mfd/Kconfig
@@ -1318,9 +1318,8 @@ config MFD_PALMAS
 
 config TPS6105X
 	tristate "TI TPS61050/61052 Boost Converters"
-	depends on I2C
+	depends on I2C && REGULATOR
 	select REGMAP_I2C
-	select REGULATOR
 	select MFD_CORE
 	select REGULATOR_FIXED_VOLTAGE
 	help
diff --git a/drivers/net/ethernet/stmicro/Kconfig b/drivers/net/ethernet/stmicro/Kconfig
index ecd7a5edef5d..259c4abf0247 100644
--- a/drivers/net/ethernet/stmicro/Kconfig
+++ b/drivers/net/ethernet/stmicro/Kconfig
@@ -5,7 +5,7 @@
 config NET_VENDOR_STMICRO
 	bool "STMicroelectronics devices"
 	default y
-	depends on HAS_IOMEM
+	depends on HAS_IOMEM && !IPIPE
 	---help---
 	  If you have a network (Ethernet) card based on Synopsys Ethernet IP
 	  Cores, say Y.
diff --git a/drivers/of/Kconfig b/drivers/of/Kconfig
index 5e1315900562..0de8e954649b 100644
--- a/drivers/of/Kconfig
+++ b/drivers/of/Kconfig
@@ -4,6 +4,7 @@ config DTC
 
 menuconfig OF
 	bool "Device Tree and Open Firmware support"
+	depends on !IPIPE
 	help
 	  This option enables the device tree infrastructure.
 	  It is automatically selected by platforms that need it or can
diff --git a/drivers/pci/Kconfig b/drivers/pci/Kconfig
index 56ff8f6d31fc..bc2cd56caf41 100644
--- a/drivers/pci/Kconfig
+++ b/drivers/pci/Kconfig
@@ -100,7 +100,7 @@ config PCI_LOCKLESS_CONFIG
 
 config PCI_IOV
 	bool "PCI IOV support"
-	depends on PCI
+	depends on PCI && !IPIPE
 	select PCI_ATS
 	help
 	  I/O Virtualization is a PCI feature supported by some devices
@@ -111,7 +111,7 @@ config PCI_IOV
 
 config PCI_PRI
 	bool "PCI PRI support"
-	depends on PCI
+	depends on PCI && !IPIPE
 	select PCI_ATS
 	help
 	  PRI is the PCI Page Request Interface. It allows PCI devices that are
@@ -121,7 +121,7 @@ config PCI_PRI
 
 config PCI_PASID
 	bool "PCI PASID support"
-	depends on PCI
+	depends on PCI && !IPIPE
 	select PCI_ATS
 	help
 	  Process Address Space Identifiers (PASIDs) can be used by PCI devices
diff --git a/drivers/pci/hotplug/Kconfig b/drivers/pci/hotplug/Kconfig
index e9f78eb390d2..c99a0435c01b 100644
--- a/drivers/pci/hotplug/Kconfig
+++ b/drivers/pci/hotplug/Kconfig
@@ -5,7 +5,7 @@
 
 menuconfig HOTPLUG_PCI
 	bool "Support for PCI Hotplug"
-	depends on PCI && SYSFS
+	depends on PCI && SYSFS && !IPIPE
 	---help---
 	  Say Y here if you have a motherboard with a PCI Hotplug controller.
 	  This allows you to add and remove PCI cards while the machine is
diff --git a/drivers/pci/pcie/Kconfig b/drivers/pci/pcie/Kconfig
index 0a1e9d379bc5..ea9e067d118c 100644
--- a/drivers/pci/pcie/Kconfig
+++ b/drivers/pci/pcie/Kconfig
@@ -4,7 +4,7 @@
 #
 config PCIEPORTBUS
 	bool "PCI Express Port Bus support"
-	depends on PCI
+	depends on PCI && !IPIPE
 	help
 	  This automatically enables PCI Express Port Bus support. Users can
 	  choose Native Hot-Plug support, Advanced Error Reporting support,
diff --git a/drivers/perf/Kconfig b/drivers/perf/Kconfig
index 08ebaf7cca8b..a3bee7d0aa88 100644
--- a/drivers/perf/Kconfig
+++ b/drivers/perf/Kconfig
@@ -3,7 +3,7 @@
 #
 
 menu "Performance monitor support"
-	depends on PERF_EVENTS
+	depends on PERF_EVENTS && !IPIPE
 
 config ARM_CCI_PMU
 	tristate "ARM CCI PMU driver"
diff --git a/drivers/pinctrl/Kconfig b/drivers/pinctrl/Kconfig
index e86752be1f19..f909c7c975b5 100644
--- a/drivers/pinctrl/Kconfig
+++ b/drivers/pinctrl/Kconfig
@@ -4,6 +4,7 @@
 
 menuconfig PINCTRL
 	bool "Pin controllers"
+	depends on !IPIPE
 
 if PINCTRL
 
diff --git a/drivers/power/avs/Kconfig b/drivers/power/avs/Kconfig
index a67eeace6a89..5d2d8bc3653d 100644
--- a/drivers/power/avs/Kconfig
+++ b/drivers/power/avs/Kconfig
@@ -1,5 +1,6 @@
 menuconfig POWER_AVS
 	bool "Adaptive Voltage Scaling class support"
+	depends on !IPIPE
 	help
 	  AVS is a power management technique which finely controls the
 	  operating voltage of a device in order to optimize (i.e. reduce)
diff --git a/drivers/power/reset/Kconfig b/drivers/power/reset/Kconfig
index 6533aa560aa1..a0483fc18363 100644
--- a/drivers/power/reset/Kconfig
+++ b/drivers/power/reset/Kconfig
@@ -1,5 +1,6 @@
 menuconfig POWER_RESET
 	bool "Board level reset or power off"
+	depends on !IPIPE
 	help
 	  Provides a number of drivers which either reset a complete board
 	  or shut it down, by manipulating the main power supply on the board.
diff --git a/drivers/powercap/Kconfig b/drivers/powercap/Kconfig
index 6ac27e5908f5..2e3c184c4b0e 100644
--- a/drivers/powercap/Kconfig
+++ b/drivers/powercap/Kconfig
@@ -4,6 +4,7 @@
 
 menuconfig POWERCAP
 	bool "Generic powercap sysfs driver"
+	depends on !IPIPE
 	help
 	  The power capping sysfs interface allows kernel subsystems to expose power
 	  capping settings to user space in a consistent way.  Usually, it consists
diff --git a/drivers/pwm/Kconfig b/drivers/pwm/Kconfig
index 504d252716f2..3f8b0bc3a060 100644
--- a/drivers/pwm/Kconfig
+++ b/drivers/pwm/Kconfig
@@ -1,5 +1,6 @@
 menuconfig PWM
 	bool "Pulse-Width Modulation (PWM) Support"
+	depends on !IPIPE
 	help
 	  Generic Pulse-Width Modulation (PWM) support.
 
diff --git a/drivers/ras/Kconfig b/drivers/ras/Kconfig
index 4c3c67d13254..3673ba89af3a 100644
--- a/drivers/ras/Kconfig
+++ b/drivers/ras/Kconfig
@@ -1,5 +1,6 @@
 menuconfig RAS
 	bool "Reliability, Availability and Serviceability (RAS) features"
+	depends on !IPIPE
 	help
 	  Reliability, availability and serviceability (RAS) is a computer
 	  hardware engineering term. Computers designed with higher levels
diff --git a/drivers/regulator/Kconfig b/drivers/regulator/Kconfig
index 329cdd33ed62..e8af9cb70bad 100644
--- a/drivers/regulator/Kconfig
+++ b/drivers/regulator/Kconfig
@@ -1,5 +1,6 @@
 menuconfig REGULATOR
 	bool "Voltage and Current Regulator Support"
+	depends on !IPIPE
 	help
 	  Generic Voltage and Current Regulator support.
 
diff --git a/drivers/remoteproc/Kconfig b/drivers/remoteproc/Kconfig
index 052d4dd347f9..cf2297fcb30e 100644
--- a/drivers/remoteproc/Kconfig
+++ b/drivers/remoteproc/Kconfig
@@ -1,4 +1,5 @@
 menu "Remoteproc drivers"
+	depends on !IPIPE
 
 config REMOTEPROC
 	tristate "Support for Remote Processor subsystem"
diff --git a/drivers/reset/Kconfig b/drivers/reset/Kconfig
index 13d28fdbdbb5..b0c789905841 100644
--- a/drivers/reset/Kconfig
+++ b/drivers/reset/Kconfig
@@ -3,6 +3,7 @@ config ARCH_HAS_RESET_CONTROLLER
 
 menuconfig RESET_CONTROLLER
 	bool "Reset Controller Support"
+	depends on !IPIPE
 	default y if ARCH_HAS_RESET_CONTROLLER
 	help
 	  Generic Reset Controller support.
diff --git a/drivers/rpmsg/Kconfig b/drivers/rpmsg/Kconfig
index d0322b41eca5..e11bc9ccc1ed 100644
--- a/drivers/rpmsg/Kconfig
+++ b/drivers/rpmsg/Kconfig
@@ -1,6 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 
 menu "Rpmsg drivers"
+	depends on !IPIPE
 
 # RPMSG always gets selected by whoever wants it
 config RPMSG
diff --git a/drivers/scsi/ufs/Kconfig b/drivers/scsi/ufs/Kconfig
index e09fe6ab3572..1da21a34a934 100644
--- a/drivers/scsi/ufs/Kconfig
+++ b/drivers/scsi/ufs/Kconfig
@@ -34,7 +34,7 @@
 
 config SCSI_UFSHCD
 	tristate "Universal Flash Storage Controller Driver Core"
-	depends on SCSI && SCSI_DMA
+	depends on SCSI && SCSI_DMA && !IPIPE
 	select PM_DEVFREQ
 	select DEVFREQ_GOV_SIMPLE_ONDEMAND
 	select NLS
diff --git a/drivers/sfi/Kconfig b/drivers/sfi/Kconfig
index dd115121e0b6..e1008c0ec50c 100644
--- a/drivers/sfi/Kconfig
+++ b/drivers/sfi/Kconfig
@@ -4,6 +4,7 @@
 
 menuconfig SFI
 	bool "SFI (Simple Firmware Interface) Support"
+	depends on !IPIPE
 	---help---
 	The Simple Firmware Interface (SFI) provides a lightweight method
 	for platform firmware to pass information to the operating system
diff --git a/drivers/soc/Kconfig b/drivers/soc/Kconfig
index c07b4a85253f..c82ed2a452d8 100644
--- a/drivers/soc/Kconfig
+++ b/drivers/soc/Kconfig
@@ -1,4 +1,5 @@
 menu "SOC (System On Chip) specific Drivers"
+	depends on !IPIPE
 
 source "drivers/soc/actions/Kconfig"
 source "drivers/soc/amlogic/Kconfig"
diff --git a/drivers/usb/chipidea/Kconfig b/drivers/usb/chipidea/Kconfig
index ee34e9046f7e..d6a85cd36b3a 100644
--- a/drivers/usb/chipidea/Kconfig
+++ b/drivers/usb/chipidea/Kconfig
@@ -1,8 +1,8 @@
 config USB_CHIPIDEA
 	tristate "ChipIdea Highspeed Dual Role Controller"
 	depends on ((USB_EHCI_HCD && USB_GADGET) || (USB_EHCI_HCD && !USB_GADGET) || (!USB_EHCI_HCD && USB_GADGET)) && HAS_DMA
+	depends on RESET_CONTROLLER
 	select EXTCON
-	select RESET_CONTROLLER
 	select USB_ULPI_BUS
 	help
 	  Say Y here if your system has a dual role high speed USB
diff --git a/drivers/virt/Kconfig b/drivers/virt/Kconfig
index 8d9cdfbd6bcc..a269093636aa 100644
--- a/drivers/virt/Kconfig
+++ b/drivers/virt/Kconfig
@@ -4,6 +4,7 @@
 
 menuconfig VIRT_DRIVERS
 	bool "Virtualization drivers"
+	depends on !IPIPE
 	---help---
 	  Say Y here to get to see options for device drivers that support
 	  virtualization environments.
diff --git a/drivers/virtio/Kconfig b/drivers/virtio/Kconfig
index 35897649c24f..e59a21e0851d 100644
--- a/drivers/virtio/Kconfig
+++ b/drivers/virtio/Kconfig
@@ -7,6 +7,7 @@ config VIRTIO
 
 menuconfig VIRTIO_MENU
 	bool "Virtio drivers"
+	depends on !IPIPE
 	default y
 
 if VIRTIO_MENU
diff --git a/drivers/watchdog/Kconfig b/drivers/watchdog/Kconfig
index 92fdc7dc2ede..53908ae1f198 100644
--- a/drivers/watchdog/Kconfig
+++ b/drivers/watchdog/Kconfig
@@ -5,6 +5,7 @@
 
 menuconfig WATCHDOG
 	bool "Watchdog Timer Support"
+	depends on !IPIPE
 	---help---
 	  If you say Y here (and to one of the following options) and create a
 	  character special file /dev/watchdog with major number 10 and minor
diff --git a/fs/Kconfig.binfmt b/fs/Kconfig.binfmt
index b795f8da81f3..46098fdc0885 100644
--- a/fs/Kconfig.binfmt
+++ b/fs/Kconfig.binfmt
@@ -159,6 +159,7 @@ config BINFMT_EM86
 
 config BINFMT_MISC
 	tristate "Kernel support for MISC binaries"
+	depends on !IPIPE
 	---help---
 	  If you say Y here, it will be possible to plug wrapper-driven binary
 	  formats into the kernel. You will like this especially when you use
diff --git a/include/linux/ipipe.h b/include/linux/ipipe.h
index fe90cb55d462..785d5860a3f1 100644
--- a/include/linux/ipipe.h
+++ b/include/linux/ipipe.h
@@ -30,9 +30,7 @@
 #include <linux/thread_info.h>
 #include <linux/ipipe_debug.h>
 #include <asm/ptrace.h>
-#ifdef CONFIG_HAVE_IPIPE_SUPPORT
 #include <asm/ipipe.h>
-#endif
 
 struct cpuidle_device;
 struct cpuidle_state;
diff --git a/init/Kconfig b/init/Kconfig
index e137ff4494dc..844c4e287976 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -69,7 +69,7 @@ config INIT_ENV_ARG_LIMIT
 
 config COMPILE_TEST
 	bool "Compile also drivers which will not load"
-	depends on HAS_IOMEM
+	depends on HAS_IOMEM && !IPIPE
 	help
 	  Some drivers can be compiled on a different platform than they are
 	  intended to be run on. Despite they cannot be loaded there (or even
@@ -254,7 +254,7 @@ config ARCH_NO_SWAP
 
 config SWAP
 	bool "Support for paging of anonymous memory (swap)"
-	depends on MMU && BLOCK && !ARCH_NO_SWAP
+	depends on MMU && BLOCK && !ARCH_NO_SWAP && !IPIPE
 	default y
 	help
 	  This option allows you to choose whether you want to have support
@@ -317,6 +317,7 @@ config CROSS_MEMORY_ATTACH
 
 config USELIB
 	bool "uselib syscall"
+	depends on !IPIPE
 	def_bool ALPHA || M68K || SPARC || X86_32 || IA32_EMULATION
 	help
 	  This option enables the uselib syscall, a system call used in the
@@ -327,7 +328,7 @@ config USELIB
 
 config AUDIT
 	bool "Auditing support"
-	depends on NET
+	depends on NET && !IPIPE
 	help
 	  Enable auditing infrastructure that can be used with another
 	  kernel subsystem, such as SELinux (which requires this for
@@ -412,6 +413,7 @@ endchoice
 config IRQ_TIME_ACCOUNTING
 	bool "Fine granularity task level IRQ time accounting"
 	depends on HAVE_IRQ_TIME_ACCOUNTING && !VIRT_CPU_ACCOUNTING_NATIVE
+	depends on !IPIPE
 	help
 	  Select this option to enable fine granularity task irq time
 	  accounting. This is done by reading a timestamp on each
@@ -761,6 +763,7 @@ config CGROUP_WRITEBACK
 
 menuconfig CGROUP_SCHED
 	bool "CPU controller"
+	depends on !IPIPE
 	default n
 	help
 	  This feature lets CPU scheduler recognize task groups and control CPU
@@ -852,7 +855,7 @@ config CGROUP_HUGETLB
 
 config CPUSETS
 	bool "Cpuset controller"
-	depends on SMP
+	depends on SMP && !IPIPE
 	help
 	  This option will let you create and manage CPUSETs which
 	  allow dynamically partitioning a system into sets of CPUs and
@@ -880,7 +883,7 @@ config CGROUP_CPUACCT
 
 config CGROUP_PERF
 	bool "Perf controller"
-	depends on PERF_EVENTS
+	depends on PERF_EVENTS && !IPIPE
 	help
 	  This option extends the perf per-cpu mode to restrict monitoring
 	  to threads which belong to the cgroup specified and run on the
@@ -992,6 +995,7 @@ config CHECKPOINT_RESTORE
 
 config SCHED_AUTOGROUP
 	bool "Automatic process group scheduling"
+	depends on !IPIPE
 	select CGROUPS
 	select CGROUP_SCHED
 	select FAIR_GROUP_SCHED
@@ -1086,6 +1090,7 @@ config CC_OPTIMIZE_FOR_PERFORMANCE
 
 config CC_OPTIMIZE_FOR_SIZE
 	bool "Optimize for size"
+	depends on !IPIPE
 	help
 	  Enabling this option will pass "-Os" instead of "-O2" to
 	  your compiler resulting in a smaller kernel.
@@ -1107,7 +1112,7 @@ config HAVE_LD_DEAD_CODE_DATA_ELIMINATION
 config LD_DEAD_CODE_DATA_ELIMINATION
 	bool "Dead code and data elimination (EXPERIMENTAL)"
 	depends on HAVE_LD_DEAD_CODE_DATA_ELIMINATION
-	depends on EXPERT
+	depends on EXPERT && !IPIPE
 	depends on !(FUNCTION_TRACER && CC_IS_GCC && GCC_VERSION < 40800)
 	depends on $(cc-option,-ffunction-sections -fdata-sections)
 	depends on $(ld-option,--gc-sections)
@@ -1161,8 +1166,6 @@ config BPF
 
 menuconfig EXPERT
 	bool "Configure standard kernel features (expert users)"
-	# Unhide debug options, to make the on-by-default options visible
-	select DEBUG_KERNEL
 	help
 	  This option allows certain base kernel options and settings
           to be disabled or tweaked. This is for specialized
@@ -1212,7 +1215,7 @@ config SYSFS_SYSCALL
 
 config SYSCTL_SYSCALL
 	bool "Sysctl syscall support" if EXPERT
-	depends on PROC_SYSCTL
+	depends on PROC_SYSCTL && !IPIPE
 	default n
 	select SYSCTL
 	---help---
@@ -1621,7 +1624,7 @@ config VM_EVENT_COUNTERS
 config SLUB_DEBUG
 	default y
 	bool "Enable SLUB debugging support" if EXPERT
-	depends on SLUB && SYSFS
+	depends on SLUB && SYSFS && !IPIPE
 	help
 	  SLUB has extensive debug support features. Disabling these can
 	  result in significant savings in code size. This also disables
@@ -1631,7 +1634,7 @@ config SLUB_DEBUG
 config SLUB_MEMCG_SYSFS_ON
 	default n
 	bool "Enable memcg SLUB sysfs support by default" if EXPERT
-	depends on SLUB && SYSFS && MEMCG
+	depends on SLUB && SYSFS && MEMCG && !IPIPE
 	help
 	  SLUB creates a directory under /sys/kernel/slab for each
 	  allocation cache to host info and debug files. If memory
@@ -1723,7 +1726,7 @@ config SLAB_FREELIST_HARDENED
 
 config SLUB_CPU_PARTIAL
 	default y
-	depends on SLUB && SMP
+	depends on SLUB && SMP && !IPIPE
 	bool "SLUB per cpu partial cache"
 	help
 	  Per cpu partial caches accellerate objects allocation and freeing
@@ -1756,6 +1759,7 @@ config MMAP_ALLOW_UNINITIALIZED
 
 config SYSTEM_DATA_VERIFICATION
 	def_bool n
+	depends on !IPIPE
 	select SYSTEM_TRUSTED_KEYRING
 	select KEYS
 	select CRYPTO
@@ -1774,6 +1778,7 @@ config SYSTEM_DATA_VERIFICATION
 
 config PROFILING
 	bool "Profiling support"
+	depends on !IPIPE
 	help
 	  Say Y here to enable the extended profiling support mechanisms used
 	  by profilers such as OProfile.
@@ -1848,6 +1853,7 @@ config MODULE_FORCE_UNLOAD
 
 config MODVERSIONS
 	bool "Module versioning support"
+	depends on !IPIPE
 	help
 	  Usually, you have to use modules compiled with your kernel.
 	  Saying Y here makes it sometimes possible to use modules
@@ -1873,7 +1879,7 @@ config MODULE_SRCVERSION_ALL
 
 config MODULE_SIG
 	bool "Module signature verification"
-	depends on MODULES
+	depends on MODULES && !IPIPE
 	select SYSTEM_DATA_VERIFICATION
 	help
 	  Check modules for valid signatures upon load: the signature
@@ -1988,7 +1994,7 @@ endchoice
 
 config TRIM_UNUSED_KSYMS
 	bool "Trim unused exported kernel symbols"
-	depends on MODULES && !UNUSED_SYMBOLS
+	depends on MODULES && !UNUSED_SYMBOLS && !IPIPE
 	help
 	  The kernel and some modules make many symbols available for
 	  other modules to use via EXPORT_SYMBOL() and variants. Depending
diff --git a/kernel/gcov/Kconfig b/kernel/gcov/Kconfig
index 1e3823fa799b..85450d4d6e13 100644
--- a/kernel/gcov/Kconfig
+++ b/kernel/gcov/Kconfig
@@ -1,4 +1,5 @@
 menu "GCOV-based kernel profiling"
+	depends on !IPIPE
 
 config GCOV_KERNEL
 	bool "Enable gcov-based kernel profiling"
diff --git a/kernel/ipipe/Kconfig b/kernel/ipipe/Kconfig
index d17edb89f1f5..d0da39b3acdf 100644
--- a/kernel/ipipe/Kconfig
+++ b/kernel/ipipe/Kconfig
@@ -1,12 +1,37 @@
 
-config HAVE_IPIPE_SUPPORT
-       depends on GENERIC_CLOCKEVENTS
-       bool
-
 config IPIPE
 	bool "Interrupt pipeline"
-	depends on HAVE_IPIPE_SUPPORT
-	default n
+	default y
+	select 64BIT
+	select ACPI
+	select AIO
+	select BUG
+	select COREDUMP
+	select DEBUG_BUGVERBOSE
+	select DOUBLEFAULT
+	select ELF_CORE
+	select EMBEDDED
+	select EXPERT
+	select FB
+	select FB_SIMPLE
+	select FRAMEBUFFER_CONSOLE
+	select HIGH_RES_TIMERS
+	select KALLSYMS
+	select MODULES
+	select MODULE_UNLOAD
+	select PANIC_ON_OOPS
+	select PCI
+	select POSIX_TIMERS
+	select PRINTK
+	select PROC_FS
+	select PROC_SYSCTL
+	select SCHED_OMIT_FRAME_POINTER
+	select SYSFS
+	select TTY
+	select VGA_CONSOLE
+	select VT
+	select X86_PM_TIMER
+	select X86_SYSFB
 	---help---
 	  Activate this option if you want the interrupt pipeline to be
 	  compiled in.
@@ -37,11 +62,3 @@ config IPIPE_TARGET_APIREV
 
 config IPIPE_HAVE_HOSTRT
        bool
-
-config IPIPE_HAVE_EAGER_FPU
-	bool
-
-if IPIPE && ARM && RAW_PRINTK && !DEBUG_LL
-comment "CAUTION: DEBUG_LL must be selected, and properly configured for"
-comment "RAW_PRINTK to work. Otherwise, you will get no output on raw_printk()"
-endif
diff --git a/kernel/ipipe/Kconfig.debug b/kernel/ipipe/Kconfig.debug
index d1894cf62d54..05a410b04d00 100644
--- a/kernel/ipipe/Kconfig.debug
+++ b/kernel/ipipe/Kconfig.debug
@@ -6,7 +6,6 @@ config IPIPE_DEBUG
 config IPIPE_DEBUG_CONTEXT
 	bool "Check for illicit cross-domain calls"
 	depends on IPIPE_DEBUG
-	default y
 	---help---
 	  Enable this feature to arm checkpoints in the kernel that
 	  verify the correct invocation context. On entry of critical
@@ -16,7 +15,6 @@ config IPIPE_DEBUG_CONTEXT
 config IPIPE_DEBUG_INTERNAL
 	bool "Enable internal debug checks"
 	depends on IPIPE_DEBUG
-	default y
 	---help---
 	  When this feature is enabled, I-pipe will perform internal
 	  consistency checks of its subsystems, e.g. on per-cpu variable
@@ -45,7 +43,6 @@ if IPIPE_TRACE
 
 config IPIPE_TRACE_ENABLE
 	bool "Enable tracing on boot"
-	default y
 	---help---
 	  Disable this option if you want to arm the tracer after booting
 	  manually ("echo 1 > /proc/ipipe/tracer/enable"). This can reduce
@@ -53,7 +50,6 @@ config IPIPE_TRACE_ENABLE
 
 config IPIPE_TRACE_MCOUNT
 	bool "Instrument function entries"
-	default y
 	select FTRACE
 	select FUNCTION_TRACER
 	---help---
@@ -65,7 +61,6 @@ config IPIPE_TRACE_MCOUNT
 
 config IPIPE_TRACE_IRQSOFF
 	bool "Trace IRQs-off times"
-	default y
 	---help---
 	  Activate this option if I-pipe shall trace the longest path
 	  with hard-IRQs switched off.
@@ -80,7 +75,6 @@ config IPIPE_TRACE_SHIFT
 
 config IPIPE_TRACE_VMALLOC
 	bool "Use vmalloc'ed trace buffer"
-	default y if EMBEDDED
 	---help---
 	  Instead of reserving static kernel data, the required buffer
 	  is allocated via vmalloc during boot-up when this option is
diff --git a/kernel/irq/Kconfig b/kernel/irq/Kconfig
index d532bf0c5a67..37ee3c8cc381 100644
--- a/kernel/irq/Kconfig
+++ b/kernel/irq/Kconfig
@@ -123,7 +123,7 @@ config SPARSE_IRQ
 
 config GENERIC_IRQ_DEBUGFS
 	bool "Expose irq internals in debugfs"
-	depends on DEBUG_FS
+	depends on DEBUG_FS && !IPIPE
 	default n
 	---help---
 
diff --git a/kernel/power/Kconfig b/kernel/power/Kconfig
index 3a6c2f87699e..f8a6fbe3f7ce 100644
--- a/kernel/power/Kconfig
+++ b/kernel/power/Kconfig
@@ -143,6 +143,7 @@ config PM_WAKELOCKS_GC
 
 config PM
 	bool "Device power management core functionality"
+	depends on !IPIPE
 	---help---
 	  Enable functionality allowing I/O devices to be put into energy-saving
 	  (low power) states, for example after a specified period of inactivity
diff --git a/kernel/rcu/Kconfig.debug b/kernel/rcu/Kconfig.debug
index 2034d02e8ee3..0ec7d1d33a14 100644
--- a/kernel/rcu/Kconfig.debug
+++ b/kernel/rcu/Kconfig.debug
@@ -5,7 +5,7 @@
 menu "RCU Debugging"
 
 config PROVE_RCU
-	def_bool PROVE_LOCKING && !IPIPE
+	def_bool PROVE_LOCKING
 
 config TORTURE_TEST
 	tristate
diff --git a/kernel/trace/Kconfig b/kernel/trace/Kconfig
index 245c6243f4f2..7a312521b2f1 100644
--- a/kernel/trace/Kconfig
+++ b/kernel/trace/Kconfig
@@ -132,6 +132,7 @@ if TRACING_SUPPORT
 
 menuconfig FTRACE
 	bool "Tracers"
+	depends on IPIPE_TRACE || !IPIPE
 	default y if DEBUG_KERNEL
 	help
 	  Enable the kernel tracing infrastructure.
diff --git a/lib/Kconfig.debug b/lib/Kconfig.debug
index 8797940742c2..0e30efb1f713 100644
--- a/lib/Kconfig.debug
+++ b/lib/Kconfig.debug
@@ -81,6 +81,7 @@ config DYNAMIC_DEBUG
 	default n
 	depends on PRINTK
 	depends on DEBUG_FS
+	depends on !IPIPE
 	help
 
 	  Compiles debug level messages into the kernel, which would not
@@ -237,7 +238,7 @@ config FRAME_WARN
 
 config STRIP_ASM_SYMS
 	bool "Strip assembler-generated symbols during link"
-	default n
+	default y
 	help
 	  Strip internal assembler-generated symbols during a link (symbols
 	  that look like '.Lxxx') so they don't pollute the output of
@@ -254,6 +255,7 @@ config READABLE_ASM
 
 config UNUSED_SYMBOLS
 	bool "Enable unused/obsolete exported symbols"
+	depends on !IPIPE
 	default y if X86
 	help
 	  Unused but exported symbols make the kernel needlessly bigger.  For
@@ -313,6 +315,7 @@ config HEADERS_CHECK
 
 config DEBUG_SECTION_MISMATCH
 	bool "Enable full Section mismatch analysis"
+	depends on !IPIPE
 	help
 	  The section mismatch analysis checks if there are illegal
 	  references from one section to another section.
@@ -351,6 +354,17 @@ config SECTION_MISMATCH_WARN_ONLY
 
 	  If unsure, say Y.
 
+config DEBUG_FORCE_FUNCTION_ALIGN_64B
+	bool "Force all function address 64B aligned" if EXPERT
+	help
+	  There are cases that a commit from one domain changes the function
+	  address alignment of other domains, and cause magic performance
+	  bump (regression or improvement). Enable this option will help to
+	  verify if the bump is caused by function alignment changes, while
+	  it will slightly increase the kernel size and affect icache usage.
+
+	  It is mainly for debug and performance tuning use.
+
 #
 # Select this config option from the architecture Kconfig, if it
 # is preferred to always offer frame pointers as a config
@@ -438,6 +452,7 @@ source "kernel/ipipe/Kconfig.debug"
 
 config DEBUG_KERNEL
 	bool "Kernel debugging"
+	depends on !IPIPE
 	help
 	  Say Y here if you are developing drivers or trying to debug and
 	  identify kernel problems.
@@ -535,7 +550,7 @@ config SLUB_DEBUG_ON
 config SLUB_STATS
 	default n
 	bool "Enable SLUB performance statistics"
-	depends on SLUB && SYSFS
+	depends on SLUB && SYSFS && !IPIPE
 	help
 	  SLUB statistics are useful to debug SLUBs allocation behavior in
 	  order find ways to optimize the allocator. This should never be
@@ -795,7 +810,15 @@ config DEBUG_SHIRQ
 	  Drivers ought to be able to handle interrupts coming in at those
 	  points; some don't and need to be caught.
 
+#
+# Enables a timestamp based low pass filter to compensate for perf based
+# hard lockup detection which runs too fast due to turbo modes.
+#
+config HARDLOCKUP_CHECK_TIMESTAMP
+	bool
+
 menu "Debug Lockups and Hangs"
+	depends on !IPIPE
 
 config LOCKUP_DETECTOR
 	bool
@@ -841,13 +864,6 @@ config HARDLOCKUP_DETECTOR_PERF
 	bool
 	select SOFTLOCKUP_DETECTOR
 
-#
-# Enables a timestamp based low pass filter to compensate for perf based
-# hard lockup detection which runs too fast due to turbo modes.
-#
-config HARDLOCKUP_CHECK_TIMESTAMP
-	bool
-
 #
 # arch/ can define HAVE_HARDLOCKUP_DETECTOR_ARCH to provide their own hard
 # lockup detector rather than the perf based detector.
@@ -1020,6 +1036,7 @@ config SCHED_STACK_END_CHECK
 
 config DEBUG_TIMEKEEPING
 	bool "Enable extra timekeeping sanity checking"
+	depends on !IPIPE
 	help
 	  This option will enable additional timekeeping sanity checks
 	  which may be helpful when diagnosing issues where timekeeping
@@ -1042,6 +1059,7 @@ config DEBUG_PREEMPT
 	  will detect preemption count underflows.
 
 menu "Lock Debugging (spinlocks, mutexes, etc...)"
+	depends on !IPIPE
 
 config LOCK_DEBUGGING_SUPPORT
 	bool
@@ -1320,7 +1338,7 @@ config HAVE_DEBUG_BUGVERBOSE
 	bool
 
 config DEBUG_BUGVERBOSE
-	bool "Verbose BUG() reporting (adds 70K)" if DEBUG_KERNEL && EXPERT
+	bool "Verbose BUG() reporting (adds 70K)" if DEBUG_KERNEL || EXPERT
 	depends on BUG && (GENERIC_BUG || HAVE_DEBUG_BUGVERBOSE)
 	default y
 	help
@@ -1622,7 +1640,7 @@ source kernel/trace/Kconfig
 
 config PROVIDE_OHCI1394_DMA_INIT
 	bool "Remote debugging over FireWire early on boot"
-	depends on PCI && X86
+	depends on PCI && X86 && !IPIPE
 	help
 	  If you want to debug problems which hang or crash the kernel early
 	  on boot and the crashing machine has a FireWire port, you can use
@@ -1651,6 +1669,7 @@ config PROVIDE_OHCI1394_DMA_INIT
 
 config DMA_API_DEBUG
 	bool "Enable debugging of DMA-API usage"
+	depends on !IPIPE
 	select NEED_DMA_MAP_STATE
 	help
 	  Enable this option to debug the use of the DMA API by device drivers.
@@ -1687,6 +1706,7 @@ config DMA_API_DEBUG_SG
 
 menuconfig RUNTIME_TESTING_MENU
 	bool "Runtime Testing"
+	depends on !IPIPE
 	def_bool y
 
 if RUNTIME_TESTING_MENU
@@ -1979,7 +1999,7 @@ endif # RUNTIME_TESTING_MENU
 
 config MEMTEST
 	bool "Memtest"
-	depends on HAVE_MEMBLOCK
+	depends on HAVE_MEMBLOCK && !IPIPE
 	---help---
 	  This option adds a kernel parameter 'memtest', which allows memtest
 	  to be set.
diff --git a/lib/Kconfig.ubsan b/lib/Kconfig.ubsan
index 98fa559ebd80..f2fd8c649091 100644
--- a/lib/Kconfig.ubsan
+++ b/lib/Kconfig.ubsan
@@ -3,6 +3,7 @@ config ARCH_HAS_UBSAN_SANITIZE_ALL
 
 config UBSAN
 	bool "Undefined behaviour sanity checker"
+	depends on !IPIPE
 	help
 	  This option enables undefined behaviour sanity checker
 	  Compile-time instrumentation is used to detect various undefined
diff --git a/lib/xz/Kconfig b/lib/xz/Kconfig
index 12d2d777f36b..06552fb30990 100644
--- a/lib/xz/Kconfig
+++ b/lib/xz/Kconfig
@@ -16,26 +16,31 @@ config XZ_DEC_X86
 config XZ_DEC_POWERPC
 	bool "PowerPC BCJ filter decoder" if EXPERT
 	default y
+	depends on PPC
 	select XZ_DEC_BCJ
 
 config XZ_DEC_IA64
 	bool "IA-64 BCJ filter decoder" if EXPERT
 	default y
+	depends on IA64
 	select XZ_DEC_BCJ
 
 config XZ_DEC_ARM
 	bool "ARM BCJ filter decoder" if EXPERT
 	default y
+	depends on ARM
 	select XZ_DEC_BCJ
 
 config XZ_DEC_ARMTHUMB
 	bool "ARM-Thumb BCJ filter decoder" if EXPERT
 	default y
+	depends on ARM_THUMB || ARM_THUMBEE
 	select XZ_DEC_BCJ
 
 config XZ_DEC_SPARC
 	bool "SPARC BCJ filter decoder" if EXPERT
 	default y
+	depends on SPARC
 	select XZ_DEC_BCJ
 
 endif
diff --git a/mm/Kconfig b/mm/Kconfig
index b457e94ae618..7dc736292380 100644
--- a/mm/Kconfig
+++ b/mm/Kconfig
@@ -523,6 +523,7 @@ config CMA_AREAS
 config MEM_SOFT_DIRTY
 	bool "Track memory changes"
 	depends on CHECKPOINT_RESTORE && HAVE_ARCH_SOFT_DIRTY && PROC_FS
+	depends on !IPIPE
 	select PROC_PAGE_MONITOR
 	help
 	  This option enables memory changes tracking by introducing a
@@ -650,7 +651,7 @@ config DEFERRED_STRUCT_PAGE_INIT
 
 config IDLE_PAGE_TRACKING
 	bool "Enable idle page tracking"
-	depends on SYSFS && MMU
+	depends on SYSFS && MMU && !IPIPE
 	select PAGE_EXTENSION if !64BIT
 	help
 	  This feature allows to estimate the amount of user pages that have
@@ -747,6 +748,7 @@ config ARCH_HAS_PKEYS
 config PERCPU_STATS
 	bool "Collect percpu memory statistics"
 	default n
+	depends on !IPIPE
 	help
 	  This feature collects and exposes statistics via debugfs. The
 	  information includes global and per chunk statistics, which can
diff --git a/mm/Kconfig.debug b/mm/Kconfig.debug
index 9a7b8b049d04..6e42ff79e150 100644
--- a/mm/Kconfig.debug
+++ b/mm/Kconfig.debug
@@ -1,5 +1,6 @@
 config PAGE_EXTENSION
 	bool "Extend memmap on extra space for more information on page"
+	depends on !IPIPE
 	---help---
 	  Extend memmap on extra space for more information on page. This
 	  could be used for debugging features that need to insert extra
@@ -41,6 +42,7 @@ config DEBUG_PAGEALLOC_ENABLE_DEFAULT
 
 config PAGE_POISONING
 	bool "Poison pages after freeing"
+	depends on !IPIPE
 	select PAGE_POISONING_NO_SANITY if HIBERNATION
 	---help---
 	  Fill the pages with poison patterns after free_pages() and verify
diff --git a/net/wireless/Kconfig b/net/wireless/Kconfig
index 41722046b937..f55211502850 100644
--- a/net/wireless/Kconfig
+++ b/net/wireless/Kconfig
@@ -89,8 +89,8 @@ config CFG80211_CERTIFICATION_ONUS
 
 config CFG80211_REQUIRE_SIGNED_REGDB
 	bool "require regdb signature" if CFG80211_CERTIFICATION_ONUS
-	default y
-	select SYSTEM_DATA_VERIFICATION
+	default n
+	depends on SYSTEM_DATA_VERIFICATION
 	help
 	  Require that in addition to the "regulatory.db" file a
 	  "regulatory.db.p7s" can be loaded with a valid PKCS#7
diff --git a/samples/Kconfig b/samples/Kconfig
index ad1ec7016d4c..9aec662a6b58 100644
--- a/samples/Kconfig
+++ b/samples/Kconfig
@@ -1,6 +1,6 @@
 menuconfig SAMPLES
 	bool "Sample kernel code"
-	depends on !UML
+	depends on !UML && !IPIPE
 	help
 	  You can build and test sample kernel code here.
 
diff --git a/security/Kconfig b/security/Kconfig
index 27e73c3b4392..8544d4052bbc 100644
--- a/security/Kconfig
+++ b/security/Kconfig
@@ -8,6 +8,7 @@ source security/keys/Kconfig
 
 config SECURITY_DMESG_RESTRICT
 	bool "Restrict unprivileged access to the kernel syslog"
+	depends on !IPIPE
 	default n
 	help
 	  This enforces restrictions on unprivileged users reading the kernel
@@ -22,6 +23,7 @@ config SECURITY
 	bool "Enable different security models"
 	depends on SYSFS
 	depends on MULTIUSER
+	depends on !IPIPE
 	help
 	  This allows you to choose different security modules to be
 	  configured into your kernel.
@@ -38,6 +40,7 @@ config SECURITY_WRITABLE_HOOKS
 
 config SECURITYFS
 	bool "Enable the securityfs filesystem"
+	depends on !IPIPE
 	help
 	  This will build the securityfs filesystem.  It is currently used by
 	  the TPM bios character driver and IMA, an integrity provider.  It is
@@ -57,7 +60,7 @@ config SECURITY_NETWORK
 config PAGE_TABLE_ISOLATION
 	bool "Remove the kernel mapping in user mode"
 	default y
-	depends on (X86_64 || X86_PAE) && !UML
+	depends on (X86_64 || X86_PAE) && !UML && !IPIPE
 	help
 	  This feature reduces the number of hardware side channels by
 	  ensuring that the majority of kernel addresses are not mapped
@@ -152,7 +155,7 @@ config HAVE_HARDENED_USERCOPY_ALLOCATOR
 
 config HARDENED_USERCOPY
 	bool "Harden memory copies between kernel and userspace"
-	depends on HAVE_HARDENED_USERCOPY_ALLOCATOR
+	depends on HAVE_HARDENED_USERCOPY_ALLOCATOR && !IPIPE
 	imply STRICT_DEVMEM
 	help
 	  This option checks for obviously wrong memory regions when
diff --git a/sound/soc/Kconfig b/sound/soc/Kconfig
index 1cf11cf51e1d..3e16909d7895 100644
--- a/sound/soc/Kconfig
+++ b/sound/soc/Kconfig
@@ -4,6 +4,7 @@
 
 menuconfig SND_SOC
 	tristate "ALSA for SoC audio support"
+	depends on !IPIPE
 	select SND_PCM
 	select AC97_BUS if SND_SOC_AC97_BUS
 	select SND_JACK
-- 
2.42.0

